### ISSUE TITLE
a "destructor" callback for reload time

### ISSUE THREAD

**Issue by yoavg** [Hammerspoon/hammerspoon#687]
It would be useful to be able to set a callback that will be called when the configuration module is reloaded, to clean-up stuff that was created during module loading. I currently add this clean-up code in my reload keybindings, but it will not trigger when using "reload config" from the hammerspoon menu.

As a specific use case: I'm using @asmagill 's undocumented spaces module to implement a scratchpad behaviour, where I can send certain windows to a hidden space and back with a hotkey. As part of that, I create the dedicated "hidden" space when the configuration loads. I now need to `spaces.remove()` it when I reload, or I will accumulate extra spaces over time.

**Comment by lowne** [comment]
+1 - user-overridable `hs.onDestroy()` would do the trick (called before luadestroystate)

**Comment by asmagill** [comment]
see latest release.

**Comment by asmagill** [comment]
(sorry, posted to wrong thread... I think I deleted my comment, so... sorry again!)

**Comment by asmagill** [comment]
tl;dr: +1 I agree, something invoked before lua_destroy() would be nice.

I think instead of a user-overridable function, though, I'd do something that lets you register/unregister a function so any module could add to the list without having to be careful not to obliterate something else's cleanup code.

Longer version:

Generally this is what the __gc metamethod, which can be attached to any table, is for.  

However, doing this correctly does require some thought because the order in which things are collected/destroyed is not guaranteed to be predictable (it's usually last created, first destroyed, but not always...), so doing anything complex, especially outside of the scope of your destructors module space, is... tricky... because it can cause things to be reloaded or their finalizers to be delayed, and most of our modules assume that a userdata's gc is only called once.  I think this was the source of some of my spurious crashes which have disappeared since I've stopped this practice.

**Comment by lowne** [comment]
@asmagill I think I get what you're saying, and what your proposed solution would look like, but your __gc critique piqued my interest; can you share an example of the sort of inter-dependency that could cause trouble without guaranteed ordering? (ideally something less convoluted than window/application/uielement)

**Comment by yoavg** [comment]
@asmagill I am really new to Lua (and even more so to OSX land) so I will leave the implementation details up to you guys ;) I do like the function-registration approach, though.

**Comment by asmagill** [comment]
The TL;DR of this comes down to 2 points: I'm guessing -- I've not reviewed the lua code for gc and lua_destroy() that closely; but observing this next point has dropped my unexplained crashes to almost 0, so I must be doing something better; and that point is this: a simple and to the point __gc which only does what it has to *and nothing else*, seems to be a lot more reliable and stable -- get fancy at your own peril.

The benefit of the hs.destroy (or whatever we call it) is that it will occur before *any* garbage collection due to a shutdown or reload occurs.  You don't have to worry at all about what is or isn't in scope or about what you might be "resurrecting" or causing to finalize multiple times.

@lowne, this is mainly conjecture based on observation and reading about some of the problems people have had with garbage collection in the lua forums, not a close scrutiny of the actual code, so let me just preface this by saying I'm not 100% convinced I'm right about the interactions -- however, when I removed or simplified the __gc methods attached to tables in my configuration files, crashes which could be traced to memory being freed by the lua_destroy function (this is invoked during garbage collection for a reload or shutdown of lua)... well, they all but disappeared.

I was using __gc methods to save settings one last time, write things to disk, manually try to shutdown/delete all of the sub-objects (timers, drawings, etc.) used by the file (ignoring the fact that their own __gc methods would do this for me, and probably in a smarter order than I could come up with), etc.  If I had to guess, I think it was that some of the __gc methods were requiring (loading) modules which weren't actually used anywhere else in the "module's" code... probably ok, when you're manually removing things (setting a table to nil, for example), and many other active things are using the same module too, so the fact that I "require" it won't change the fact that it was going to stay around anyways.  But when everything is being released at once, and here I go adding new ones...  At best, it changes the order in which a given module is ultimately released... at worst it causes its __gc method (or methods if it creates/maintains multiple objects) to be called more than once, which may not be possible in the final steps of lua_destroy().

Most of our compiled __gc methods should be pretty safe -- they either check that their data is still valid (not nil or an invalid reference) before actually releasing it, or we remove the metatable (and thus the __gc method) so even if the object does get collected twice, it no longer has the __gc function attached to it.  It probably wouldn't hurt to do a formal audit of this, but we've caught or corrected most of them already.

**Comment by cmsj** [comment]
So, one issue that jumps out immediately at me is that any kind of user controlled restart hook should probably be time limited and then we have the question of how to decide what that time should be.

I'm not enormously comfortable with this idea yet - I see the uses of it, but I'm worried that it's going to encourage people to either do things that are unnecessary, or risky. At this point I'd almost prefer to be able to let users reconfigure the "Reload Config" menu item to call a Lua function of their choosing.

**Comment by asmagill** [comment]
@cmsj, while I share your concerns, I'm still in favor of having a way to execute something before actual garbage collection occurs but after a user has effectively said "yes, I want to restart/quit".  I really believe that my problems primarily occurred because I was attempting to do too much of my own cleanup (last minute writing to a file, etc.) when things were literally disappearing around my code and I had no way to differentiate between what had been marked for finalization, what had already been finalized, and what hadn't even been touched yet.

It's really not that different than an init.lua that includes `while true do end`... you enter a loop and you have to force quit to get out of it (or use a debug.sethook function and hope the loop isn't in a c-function).  We can't fully guard against stupid mistakes without drastically reducing what Hammerspoon and lua can do or are.

Maybe have it disabled by default and require the use of a "hidden" setting like disabling autoload or changing your config directory path.  That would at least discourage using such a hook in a production module while still allowing a user to do so for a specific personal application.

**Comment by lowne** [comment]
To clarify the problem space, here's what I think are the different approaches that have been proposed:
1. *modules* (including pure Lua ones, and including non-HS modules provided by the user) do their own cleanup in __gc as per usual; a special, single unload hook is provided for the user *config* (because it's simpler than forcing users to wrap their configs in a "fake" table and set its metatable's __gc)
2. (by @asmagill) have a global subscriber list for the unload event that can be used by the config *and* the modules (even, in theory, HS modules) and that *maintains* order (sort of manual memory management) as opposed to the difficult-to-predict ordering of all the __gcs 

I'm trying to understand what circumstances might make option 2 necessary:
> ignoring the fact that their own __gc methods would do this for me

Yes, each module's __gc should only cleanup after itself and trust its "children"'s __gcs to be called at the right time; if this is made sure of, then, it would seem, problems can only arise with particular cases of mutual dependencies (i.e. cyclic graph) where 1. it's not guaranteed which __gc will be called first and 2. the order in which __gcs are called matters. My question (assuming the former as true) is: what are examples of the latter?

> some of the __gc methods were requiring (loading) modules

Prima facie I'd think that should never be the case, if the "mind your own garbage" rule above is followed - am I missing something?

@cmsj fwiw I also share your concerns but agree with @asmagill's counterpoint; no need to worry about time limiting the unload hook when we can freely shoot ourselves in the foot with infinite, or just really long, loops anywhere else in the config.

**Comment by asmagill** [comment]
Let me just clarify that I don't have *proof* that moving things out of some of my configuration files __gc methods reduced the crashes... I just have the observation that crashes in which the offending function was identified as lua_destroy(), which is part of a reload or quit's final garbage collection/memory release, have all but disappeared.  It's conceivable that other bug fixes affected this as well, but my experience is that most other bug fixes mostly addressed things that caused other traceback locations to be identified, not lua-destroy.

Really, the primary use case I can come up with for the way I do things is for writing out status/history/state information, either with hs.settings or to a file, when restarting.  This can also be done (as I've started to do) as things change or on a timer, but with some impact to load (many small writes, rather than one larger final one) or a potential loss of data if changes occur between the last timer invocation, and the restart.  The other things I was doing in my garbage collection methods really should be left to the internal mechanisms, but I "wanted to be sure" and "clean up after myself"... pre-ARC and pre modern GC thinking, but its how I learned.

So, I like the idea, but I'm not waiting for it...

I do think, however, that if it's offered, it should be a "register your cleanup function" approach, though, because if it's an "override this function if you want it" approach, then only one function can be registered at a time and if it doesn't save and chain to the previous value, one bad apple can break everything else that relies on it.

**Comment by cmsj** [comment]
For the registry approach we could just expose a table :D

**Comment by lowne** [comment]
If a pre-gc global hook for all the modules (option 2) isn't needed, then there's no point in having a registry, because with option 1 there'd be only *one* callback at most - the user config unload callback, which isn't even related to "garbage" at all but it's just about persistence, such as saving settings, or deleting extra spaces (which persist) as per OP.
If option 2 *is* needed (but I'm really not convinced), then the registry needs to discard multiple registrations and maintain order, i.e. a set and a list; something like (very roughly):
```lua
local unloadIdx,unloadFn=setmetatable({},{__mode='kv'}),setmetatable({},{__mode='kv'})
function hs.subscribeUnload(fn)
  if unloadIdx[fn] then return end
  table.insert(unloadFn,fn)
  unloadIdx[fn]=#unloadFn
end
function hs.onUnload()
  for i=#unloadFn,1 do unloadFn[i]() end
end
```
*However*, we'd basically be doing a bad job at replicating the garbage collector *and* burdening our code with manual memory management, so I think we should avoid this path unless we're sure that we need it and we understand exactly *why* we need it (i.e. what's "broken" with the regular __gc route).

@asmagill for non-cleanup tasks in *modules*, such as persisting settings (while avoiding "accidentally" `require`ing `hs.settings` within the __gc), a way to make absolutely sure that we still have all the needed references to can be used within __gc is to have them in the table itself:
```lua
-- module.lua
module={hssettings=require'hs.settings', somekey='module.key'}
-- ...
return setmetatable(module,{__gc=function(t)t.hssettings.set(t.somekey,t.somevalue) end})
```

**Comment by asmagill** [comment]
you would think so, but (again this is purely anecdotal with no proof but observation) not sufficient in my experience.

**Comment by cmsj** [comment]
Ok, so let's start with a single shutdown callback and we can expand from there if we ever need to.

**Comment by cmsj** [comment]
This should be done now. I took the opportunity to rework the Lua lifecycle handling in MJLua.m, so fingers crossed I haven't ruined anything :)

**Comment by cmsj** [comment]
(specifically, `hs.shutdownCallback` will be available in 0.9.44)
